libraries/AP_RCProtocol/AP_RCProtocol.h

```diff
@@ -16,45 +16,70 @@
  */
 #pragma once
 #include <AP_HAL/AP_HAL.h>
-#include <AP_Common/AP_Common.h>
 
-#define MAX_RCIN_CHANNELS 18
-#define MIN_RCIN_CHANNELS  5
+class AP_RCProtocol_Backend;
+
+#include "AP_RCProtocol_config.h"
 
-#ifndef AP_RCPROTOCOL_FASTSBUS_ENABLED
-  #ifdef IOMCU_FW
-    #define AP_RCPROTOCOL_FASTSBUS_ENABLED 0
-  #else
-    #define AP_RCPROTOCOL_FASTSBUS_ENABLED 1
-  #endif
+enum class rcprotocol_t {
+#if AP_RCPROTOCOL_PPMSUM_ENABLED || HAL_WITH_IO_MCU
+    PPM        =  0,
+#endif
+#if AP_RCPROTOCOL_IBUS_ENABLED || HAL_WITH_IO_MCU
+    IBUS       =  1,
+#endif
+#if AP_RCPROTOCOL_SBUS_ENABLED || HAL_WITH_IO_MCU
+    SBUS       =  2,
+#endif
+#if AP_RCPROTOCOL_SBUS_NI_ENABLED || HAL_WITH_IO_MCU
+    SBUS_NI    =  3,
+#endif
+#if AP_RCPROTOCOL_DSM_ENABLED || HAL_WITH_IO_MCU
```

Comment:
Doesn't this lead to weird results in the protocol array if one is missing? I thought these needed to be contiguous?

Reply:
No, you just end up with nullptrs in the array.

I'm experimenting with using an `AC_RCProtcol_Backend *_current_backend` to reduce some of the weirdnesses and keep it tightly packed, but I've got a feeling if I push for making this a compact array we'll end up with a bit more flash as each backend will need to know its protocol number for checking the bitmasks for enablement and whatnot.

Reply:
This was a different PR